Faster downloader for historical eth transfers

dshulyak opened the pull request:
Implemented on top of https://github.com/status-im/status-go/pull/1473

dshulyak committed:
Implement faster downloader for historical eth transfers

ghost commented:
<!--prchecklist--> 
### Pull Request Checklist
- [ ] Have you updated the documentation, if impacted (e.g. [docs.status.im](https://docs.status.im/docs/build_status_go.html))?

status-im-auto commented:
### Jenkins Builds
<details>
<summary>Click to see older builds (18)</summary>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | ff97f1a7 | [#1](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1475/1/) | 2019-05-31 10:03:14 | ~1 min | `linux` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-190531-100205-ff97f1.zip) |
| :heavy_check_mark: | ff97f1a7 | [#1](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1475/1/) | 2019-05-31 10:05:42 | ~3 min | `ios` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-190531-100204-ff97f1.zip) |
| :heavy_check_mark: | ff97f1a7 | [#1](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1475/1/) | 2019-05-31 10:12:27 | ~10 min | `android` | [:package: aar](https://status-go.ams3.digitaloceanspaces.com/status-go-android-190531-100201-ff97f1.aar) |
| | | | | | | |
| :heavy_check_mark: | 6ce76877 | [#2](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1475/2/) | 2019-06-03 07:29:58 | ~1 min | `linux` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-190603-072855-6ce768.zip) |
| :heavy_check_mark: | 6ce76877 | [#2](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1475/2/) | 2019-06-03 07:32:41 | ~3 min | `ios` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-190603-072854-6ce768.zip) |
| :heavy_check_mark: | 6ce76877 | [#2](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1475/2/) | 2019-06-03 07:35:21 | ~6 min | `android` | [:package: aar](https://status-go.ams3.digitaloceanspaces.com/status-go-android-190603-072852-6ce768.aar) |
| | | | | | | |
| :heavy_check_mark: | 5f659f1b | [#3](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1475/3/) | 2019-06-03 08:34:19 | ~31 sec | `linux` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-190603-083347-5f659f.zip) |
| :heavy_check_mark: | 5f659f1b | [#3](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1475/3/) | 2019-06-03 08:37:39 | ~3 min | `ios` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-190603-083346-5f659f.zip) |
| :heavy_check_mark: | 5f659f1b | [#3](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1475/3/) | 2019-06-03 08:40:23 | ~6 min | `android` | [:package: aar](https://status-go.ams3.digitaloceanspaces.com/status-go-android-190603-083345-5f659f.aar) |
| | | | | | | |
| :heavy_check_mark: | e386a005 | [#4](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1475/4/) | 2019-06-03 08:36:09 | ~34 sec | `linux` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-190603-083535-e386a0.zip) |
| :heavy_check_mark: | e386a005 | [#4](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1475/4/) | 2019-06-03 08:40:29 | ~2 min | `ios` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-190603-083740-e386a0.zip) |
| :heavy_check_mark: | e386a005 | [#4](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1475/4/) | 2019-06-03 08:46:47 | ~6 min | `android` | [:package: aar](https://status-go.ams3.digitaloceanspaces.com/status-go-android-190603-084023-e386a0.aar) |
| | | | | | | |
| :heavy_check_mark: | 79a89c11 | [#5](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1475/5/) | 2019-06-03 10:17:10 | ~32 sec | `linux` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-190603-101637-79a89c.zip) |
| :heavy_check_mark: | 79a89c11 | [#5](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1475/5/) | 2019-06-03 10:19:39 | ~3 min | `ios` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-190603-101636-79a89c.zip) |
| :heavy_check_mark: | 79a89c11 | [#5](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1475/5/) | 2019-06-03 10:23:31 | ~6 min | `android` | [:package: aar](https://status-go.ams3.digitaloceanspaces.com/status-go-android-190603-101633-79a89c.aar) |
| | | | | | | |
| :heavy_check_mark: | a2e26c87 | [#6](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1475/6/) | 2019-06-04 07:29:50 | ~49 sec | `linux` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-190604-072900-a2e26c.zip) |
| :heavy_check_mark: | a2e26c87 | [#6](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1475/6/) | 2019-06-04 07:31:48 | ~2 min | `ios` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-190604-072859-a2e26c.zip) |
| :heavy_check_mark: | a2e26c87 | [#6](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1475/6/) | 2019-06-04 07:35:02 | ~6 min | `android` | [:package: aar](https://status-go.ams3.digitaloceanspaces.com/status-go-android-190604-072858-a2e26c.aar) |
</details>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 4bd68a36 | [#7](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1475/7/) | 2019-06-04 07:32:38 | ~32 sec | `linux` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-190604-073205-4bd68a.zip) |
| :heavy_check_mark: | 4bd68a36 | [#7](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1475/7/) | 2019-06-04 07:36:03 | ~3 min | `ios` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-190604-073203-4bd68a.zip) |
| :heavy_check_mark: | 4bd68a36 | [#7](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1475/7/) | 2019-06-04 07:41:13 | ~6 min | `android` | [:package: aar](https://status-go.ams3.digitaloceanspaces.com/status-go-android-190604-073503-4bd68a.aar) |
| | | | | | | |
| :x: | e1b1d918 | [#8](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1475/8/) | 2019-06-13 06:26:50 | ~10 sec | `android` | [:page_facing_up: log](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1475/8/consoleText) |
| :x: | e1b1d918 | [#8](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1475/8/) | 2019-06-13 06:26:50 | ~8.6 sec | `linux` | [:page_facing_up: log](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1475/8/consoleText) |
| :x: | e1b1d918 | [#8](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1475/8/) | 2019-06-13 06:26:58 | ~17 sec | `ios` | [:page_facing_up: log](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1475/8/consoleText) |

dshulyak committed:
Test binary iterative downloader

dshulyak committed:
Remove unused ethbatchSize

dshulyak commented:
@yenda you can switch to this change now. it is rebased an includes last discussed changes.

yenda commented:
Here is a report for the logout bug:

https://gist.github.com/yenda/04c77cf8ba0504237918f8ffdaba4016

dshulyak committed:
Use parent ctx in iterators and hash of tx as a unique identifier

dshulyak committed:
Replace binary iterative downloader with concurrent downloader for eth

dshulyak committed:
Refactor logic from reactor

dshulyak committed:
Rework downloaders so that can accept multiple accounts

dshulyak committed:
Support for downloading transfers for multiple accounts

dshulyak committed:
Update readme with minor fixes

dshulyak committed:
Refactor issues reported by linter

dshulyak committed:
Rewrite concurrent downloader and make erc20 downloader concurrent

dshulyak committed:
Generate unique ID for every transfer

dshulyak committed:
Write better tests for reorg handling and improve handling description

dshulyak committed:
Renaming leftover

dshulyak committed:
Add logs and simplify lastKnownHeader

dshulyak committed:
Update accounts with all blocks in the set

dshulyak committed:
Encrypt database with same password taht is used for pfs db

dshulyak committed:
Rename ConcurrentRunner to AtomicGroup

dshulyak committed:
Add fast catchup after app was offline for longer than 15 blocks

When downloader is started it will go through following procedure:
- verify that the last synced block is still in canonical chain
- run fast indexer for every account from last synced block to head of the chain
- start watching block by block

dshulyak committed:
Remove ununes isSequence

dshulyak committed:
Increase timeout in test: bootstrap procedure may take longer now

dshulyak committed:
Remove unused code in database and adjust documentation
